File: api/src/models/track.rs
use crate::{models::generic::Pageable, utils::de::de_opt_i64};
use serde::{Deserialize, Serialize};

use crate::models::{album::AlbumWithArtist, artist::Artist};

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Track {
    pub id: i64,
    pub track_number: i64,
    pub title: String,
    pub file_path: String,
    pub duration_ms: i64,
    pub has_lyrics: bool,
    pub album: AlbumWithArtist,
    pub artist: Artist,
}

impl Track {
    pub fn relative_file_path(&self) -> String {
        if self.file_path.starts_with('/') {
            self.file_path[1..].into()
        } else {
            self.file_path.clone()
        }
    }
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TracksFilters {
    #[serde(default, deserialize_with = "de_opt_i64")]
    pub artist_id: Option<i64>,
    #[serde(default, deserialize_with = "de_opt_i64")]
    pub album_id: Option<i64>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TracksQuery {
    #[serde(flatten)]
    pub pageable: Pageable,
    #[serde(flatten)]
    pub filters: TracksFilters,
}

File: web/src/domain/track.ts
import z from "zod";
import type { Album } from "./album";
import type { Artist } from "./artist";
import { pageableSchema } from "./generic";

export interface Track {
  id: number;
  trackNumber: number;
  title: string;
  hasLyrics: boolean;
  artist: Artist;
  album: Album;
}

export const trackSearchSchema = pageableSchema.extend({
  artistId: z.int().min(1).optional(),
  albumId: z.int().min(1).optional(),
});

export type TrackSearch = z.infer<typeof trackSearchSchema>;
